README.md
# nsmp_barcode_app

Встраиваемое приложение для печати штрихкодов для Naumen SMP 4 (Naumen Service Desk)
и сервиса ITSM 365.

За инструкцией по настройке обращайтесь в поддержку [ITSM 365](mailto:cs@itsm365.com)
или [Naumen](mailto:supports@naumen.ru).

Распространяется под BSD
[лицензией](https://github.com/dart-lang/stagehand/blob/master/LICENSE).


CHANGELOG.md
## 1.0.0

- Готово к использованию на стандартной объектной модели ITSM 365/NSD ППВ

## 1.0.1

- Исключил из зависимостей Dartson
- Добавил читаемые ошибки при проблемах с кодированием штрихкода

## 1.0.2

- Временно исключил из зависимостей dart_barcode
- Обновил читаемые ошибки при проблемах с кодированием штрихкода

web/main.dart
import 'dart:html';

import 'package:naumen_smp_jsapi/naumen_smp_jsapi.dart';
//import 'package:dart_barcode/dart_barcode.dart';

import 'package:nsmp_barcode_app/src/nsmp/nsmp_objects.dart';
import 'package:nsmp_barcode_app/src/dart_barcode/barcode.dart';
import 'package:nsmp_barcode_app/src/dart_barcode/options.dart';

String noAssetError =
    'Не удалось получить карточку актива. Обратитесь к администратору сервера.';
String noPrintTemplateError =
    'Не удалось получить настройки печати. Обратитесь к администратору сервера.';

Map<String, String> barcodesInfo = {
  'CODE128': '<h4>Для CODE128 допустимы</h4><p>'
      'Символы пунктуации: !\"#\$%&\'()\*+,-./:;<=>?@[]^_<br/>'
      'Цифры: 0..9<br/>'
      'Заглавные буквы: A..Z</p>',
  'CODE128A': '<h4>Для CODE128A допустимы</h4><p>'
      'Символы пунктуации: !\"#\$%&\'()\*+,-./:;<=>?@[]^_<br/>'
      'Цифры: 0..9<br/>'
      'Заглавные буквы: A..Z<br/>'
      'Управляющие коды ASCII с шестнадцатиричными кодами 0H..20H</p>',
  'CODE128B': '<h4>Для CODE128B допустимы</h4><p>'
      'Символы пунктуации: !\"#\$%&\'()\*+,-./:;<=>?@[]^_<br/>'
      'Дополнительные символы: `{|}~<br/>'
      'Цифры: 0..9<br/>'
      'Заглавные буквы: A..Z<br/>'
      'Малые буквы: a..z</p>',
  'CODE128C': '<h4>Для CODE128C допустимы</h4><p>'
      'Цифры: 0..9</p>',
  'CODE39': '<h4>Для CODE39 допустимы</h4><p>'
      'Цифры: 0..9<br/>'
      'Дополнительные символы: - .\$/+% и пробел</p>',
  'EAN13': '<h4>Для EAN13 допустимы</h4><p>'
      'Цифры: 0..9</p>',
  'EAN8': '<h4>Для EAN8 допустимы</h4><p>'
      'Цифры: 0..9</p>',
  'EAN5': '<h4>Для EAN5 допустимы</h4><p>'
      'Цифры: 0..9</p>',
  'EAN2': '<h4>Для EAN2 допустимы</h4><p>'
      'Цифры: 0..9</p>',
  'UPC': '<h4>Для UPC допустимы</h4><p>'
      'Цифры: 0..9</p>',
  'ITF14': '<h4>Для ITF14 допустимы</h4><p>'
      'Цифры: 0..9</p>',
  'MSI': '<h4>Для MSI допустимы</h4><p>'
      'Цифры: 0..9</p>',
  'MSI10': '<h4>Для MSI10 допустимы</h4><p>'
      'Цифры: 0..9</p>',
  'MSI11': '<h4>Для MSI11 допустимы</h4><p>'
      'Цифры: 0..9</p>',
  'MSI1010': '<h4>Для MSI1010 допустимы</h4><p>'
      'Цифры: 0..9</p>',
  'MSI1110': '<h4>Для MSI1110 допустимы</h4><p>'
      'Цифры: 0..9</p>',
  'CODABAR': '<h4>Для CODABAR допустимы</h4><p>'
      'Цифры: 0..9</p>'
      'Специальные символы: –\$:/+.</p>',
  'PHARMACODE': '<h4>Для pharmacode допустимы</h4><p>'
      'Цифры: от 3 до 131070</p>',
};

void main() async {
  Asset asset;
  PrintTemplate printTemplate;
  String currentUid = SmpAPI.currentUUID;
  DivElement container = querySelector('#barcodeContainer');
  ImageElement imageElement = querySelector('#loading');
  try {
    asset = Asset(await SmpAPI.get(currentUid));
  } catch (e) {
    imageElement.style.display = 'none';
    container.appendText(noAssetError);
    return;
  }
  try {
    printTemplate =
        PrintTemplate.fromMap(await SmpAPI.get(asset.getPrintTemplateId()));
  } catch (e) {
    imageElement.style.display = 'none';
    container.appendText(noPrintTemplateError);
    return;
  }
  container.style.width = printTemplate.getWidth();
  container.style.height = printTemplate.getHeight();
  Element svg = querySelector('#barcode');
  Options options = new Options(
    format: printTemplate.barcode,
    displayValue: false,
    width: 1,
    height: 50,
  );
  String barcodeValue = asset.getStringProperty(printTemplate.barcodeAttr);
  imageElement.style.display = 'none';
  try {
    new DartBarcode("#barcode", barcodeValue, options);
    svg.style.removeProperty('display');
    svg.style.maxWidth = printTemplate.getWidth();
    for (String item in asset.getProperties(printTemplate.getAdditional())) {
      container.appendHtml('<div>$item</div>');
    }
    Element printAction = querySelector('#print');
    printAction.style.removeProperty('display');
    printAction.onClick.listen((_) => printJs(printTemplate, printAction));
  } catch (e) {
    DivElement barcodeInfo = new DivElement();
    HeadingElement errorTitle = new HeadingElement.h4();
    errorTitle.text = 'Ошибка подготовки штрихкода для \"$barcodeValue\"';
    String barcodeInfoText = barcodesInfo[printTemplate.barcode.toUpperCase()];
    if (barcodeInfoText == null) {
      barcodeInfoText =
          'Выбранный  метод кодирования <b>${printTemplate.barcode.toUpperCase()}</b> не поддерживается';
    }
    barcodeInfo.appendHtml(barcodeInfoText);
    container.append(errorTitle);
    container.append(barcodeInfo);
  }
}

void printJs(PrintTemplate printTemplate, Element action) {
  action.style.display = 'none';
  String bodyPadding = document.body.style.padding;
  document.body.style.padding = '0';
  document.body.style.width = printTemplate.width;
  window.print();
  action.style.removeProperty('display');
  document.body.style.padding = bodyPadding;
}


lib/src/nsmp/nsmp_rest.dart
import 'dart:async';
import 'dart:convert';
import 'package:http/browser_client.dart';
import 'package:http/http.dart';

class NsmpRest {
  static final _headers = {'Content-Type': 'application/json'};
  static String _find = '../services/rest/find';
  static String _get = '../services/rest/get';
  static String _edit = '../services/rest/edit';
  static String _create = '../services/rest/create-m2m';
  static BrowserClient _http = new BrowserClient();

  static Future<Map> findFirst(String url) async {
    try {
      List data = await find(url);
      return data.length > 0 ? data.first : {};
    } catch (e) {
      print(e.toString());
      return {};
    }
  }
  static Future<Map> getObjectByUrl(String url) async {
    try {
      final response = await _http.get(url);
      return _extractData(response);
    } catch (e) {
      return {};
    }
  }
  static Future<Map> get(String url) async {
    try {
      final response = await _http.get('$_get/$url');
      return _extractData(response);
    } catch (e) {
      return {};
    }
  }

  static Future<List> find(String url) async {
    try {
      final response = await _http.get('$_find$url');
      return _extractData(response);
    } catch (e) {
      return [];
    }
  }

  static Future<Map> create(String url, Map data) async {
    try {
      final response = await _http.post(
          '$_create$url',
          headers: _headers,
          body: JSON.encode(data)
      );
      return _extractData(response);
    } catch (e) {
      print(e.toString());
      return {};
    }
  }

  static Future<String> edit(String url, Map data) async {
    String body = JSON.encode(data);
    try {
      final response = await _http.post(
          '$_edit$url',
          headers: _headers,
          body: body
      );
      return response.body;
    } catch (e) {
      print(e.toString());
      return 'Rest error';
    }
  }

  static _extractData(Response resp) => JSON.decode(resp.body);
}

lib/src/nsmp/nsmp_objects.dart
class Asset {
  Map<String, Object> _properties;

  String getProperty(String code) => _properties[code];

  String getStringProperty(String code) => _properties[code].toString();

  List<String> getProperties(List<String> codes) =>
      codes.map(getProperty).toList();

  List<String> getStringProperties(List<String> codes) =>
      codes.map(getStringProperty).toList();

  String getPrintTemplateId({String printTemplate = 'printConfig'}) {
    Map templateParams = _properties[printTemplate];
    if(templateParams != null) {
      return templateParams['UUID'];
    }
    return null;
  }

  Asset(this._properties);

}

class PrintTemplate {
  String height;
  String width;
  String unit;
  String barcode;
  String barcodeAttr;
  String additional;
  Map<String,String> styles;

  String getHeight() => '$height$unit';
  String getWidth() => '$width$unit';
  Map<String,String> getStyles() {
    return <String,String>{
      'max-width':'$width$unit',
      'width':'$width$unit',
      'min-height':'$height$unit',
    };
  }
  List<String> getAdditional(){
    if(additional != null){
      return additional.split(new RegExp(r",\s?"));
    }
    return [];
  }

  PrintTemplate.fromMap(Map<String, Object> data) {
    this.height = data['height'];
    this.width = data['width'];
    this.unit = data['unit'];
    this.barcode = data['barcode'];
    this.barcodeAttr = data['barcodeAttr'];
    this.additional = data['additional'];
  }
}


lib/src/asset/asset_service.dart
import 'dart:async';
import 'dart:js';

import 'package:angular/core.dart';

import '../nsmp/nsmp_objects.dart';
import '../nsmp/nsmp_rest.dart';

@Injectable()
class AssetService {
  Future<Asset> getAsset(String uuid) async {
    Map response = await NsmpRest.get(uuid);
    return new Asset(response);
  }

  Future<Asset> getMockAsset() async {
    return new Asset({
      'inventoryNumbe': '000000096',
      'serialNumber': 'SN000000',
      'model': 'EPSON',
      'UUID': 'objectBase',
      'printConfig': {
        'width': '56',
        'height': '58',
        'unit': 'mm',
        'barcode': 'code39',
        'barcodeAttr': 'inventoryNumbe',
        'additional': 'inventoryNumbe,model,serialNumber'
      }
    });
  }

  Future<Asset> getCurrentAsset() async {
    String uuid = getSourceUUID();
    if (uuid != null) {
      Map response = await NsmpRest.get(uuid);
      return new Asset(response);
    }
    return null;
  }

  Future<PrintTemplate> getPrintTemplate(String uuid) async {
    if (uuid != null) {
      Map response = await NsmpRest.get(uuid);
      return new PrintTemplate.fromMap(response);
    }
    return null;
  }

  String getSourceUUID() {
    RegExp exp = new RegExp(r"#uuid:(objectBase\$\d+)");
    Match match = exp.firstMatch(context['top']['location']['hash']);
    if (match != null) {
      return match.group(1);
    }
    return null;
  }
}


lib/src/printed_card/printed_card.dart
import 'dart:async';
import 'dart:html';

import 'package:angular/angular.dart';
import 'package:angular_components/angular_components.dart';

import '../nsmp/nsmp_objects.dart';
import '../asset/asset_service.dart';
import '../asset/asset_card.dart';

@Component(
  selector: 'printed-card',
  styleUrls: const ['printed_card.css'],
  templateUrl: 'printed_card.html',
  directives: const [
    CORE_DIRECTIVES,
    materialDirectives,
    AssetCard
  ],
  providers: const [AssetService],
)
class PrintedCard implements OnInit {
  final AssetService assetService;
  PrintedCard(this.assetService);
  Asset asset;
  PrintTemplate printTemplate;
//  AssetCard assetCard;

  @override
  Future<Null> ngOnInit() async {
    asset = await assetService.getCurrentAsset();
    printTemplate = await assetService.getPrintTemplate(asset.getPrintTemplateId());
  }

  void printJs() {
    Element action = querySelector('#printed-action');
    action.style.display = 'none';
    String bodyPadding = document.body.style.padding;
    document.body.style.padding = '0';
    document.body.style.width = printTemplate.width;
    window.print();
    action.style.display = 'block';
    document.body.style.padding = bodyPadding;
  }
}


lib/src/asset/asset_card.dart
import 'dart:html';

import 'package:angular/angular.dart';
import 'package:angular_components/angular_components.dart';
import 'package:dart_barcode/dart_barcode.dart';

import '../nsmp/nsmp_objects.dart';

Map<String, String> barcodesInfo = {
  'CODE128': '<h3>Для CODE128 допустимы</h3><ol>'
      '<li>Символы пунктуации: !\"#\$%&\'()\*+,-./:;<=>?@[]^_</li>'
      '<li>Цифры: 0..9</li>'
      '<li>Заглавные буквы: A..Z</li>'
      '</ol>',
  'CODE128A': '<h3>Для CODE128A допустимы</h3><ol>'
      '<li>Символы пунктуации: !\"#\$%&\'()\*+,-./:;<=>?@[]^_</li>'
      '<li>Цифры: 0..9</li>'
      '<li>Заглавные буквы: A..Z</li>'
      '<li>Управляющие коды ASCII с шестнадцатиричными кодами 0H..20H</li>'
      '</ol>',
  'CODE128B': '<h3>Для CODE128B допустимы</h3><ol>'
      '<li>Символы пунктуации: !\"#\$%&\'()\*+,-./:;<=>?@[]^_</li>'
      '<li>Дополнительные символы: `{|}~</li>'
      '<li>Цифры: 0..9</li>'
      '<li>Заглавные буквы: A..Z</li>'
      '<li>Малые буквы: a..z</li>'
      '</ol>',
  'CODE128C': '<h3>Для CODE128C допустимы</h3><ol>'
      '<li>Цифры: 0..9</li>'
      '</ol>',
  'CODE39': '<h3>Для CODE39 допустимы</h3><ol>'
      '<li>Цифры: 0..9</li>'
      '<li>Дополнительные символы: -, ., \$, /, +, %, и пробел</li>'
      '</ol>',
  'EAN13': '<h3>Для EAN13 допустимы</h3><ol>'
      '<li>Цифры: 0..9</li>'
      '</ol>',
  'EAN8': '<h3>Для EAN8 допустимы</h3><ol>'
      '<li>Цифры: 0..9</li>'
      '</ol>',
  'EAN5': '<h3>Для EAN5 допустимы</h3><ol>'
      '<li>Цифры: 0..9</li>'
      '</ol>',
  'UPC': '<h3>Для UPC допустимы</h3><ol>'
      '<li>Цифры: 0..9</li>'
      '</ol>',
  'MSI': '<h3>Для MSI допустимы</h3><ol>'
      '<li>Цифры: 0..9</li>'
      '</ol>',
  'MSI10': '<h3>Для MSI10 допустимы</h3><ol>'
      '<li>Цифры: 0..9</li>'
      '</ol>',
  'MSI11': '<h3>Для MSI11 допустимы</h3><ol>'
      '<li>Цифры: 0..9</li>'
      '</ol>',
  'MSI1010': '<h3>Для MSI1010 допустимы</h3><ol>'
      '<li>Цифры: 0..9</li>'
      '</ol>',
  'MSI1110': '<h3>Для MSI1110 допустимы</h3><ol>'
      '<li>Цифры: 0..9</li>'
      '</ol>',
  'pharmacode': '<h3>Для pharmacode допустимы</h3><ol>'
      '<li>Цифры: от 3 до 131070</li>'
      '</ol>',

};

@Component(
    selector: 'asset-card',
    styleUrls: const ['asset_card.css'],
    templateUrl: 'asset_card.html',
    directives: const [CORE_DIRECTIVES, materialDirectives])
class AssetCard implements OnInit {
  @Input('asset')
  Asset asset;

  @Input('barcodeAttr')
  String barcodeAttr;

  @Input('barcode')
  String barcode;

  @Input('additional')
  List<String> additional;

  @override
  void ngOnInit() {
    Options options = new Options(
      format: barcode,
      displayValue: false,
      width: 1,
      height: 50,
    );
    String barcodeValue = asset.getStringProperty(barcodeAttr);
    try {
      new DartBarcode(
          "#barcode", barcodeValue, options);
    } catch (e) {
      Element error = querySelector('#error');
      DivElement barcodeInfo = new DivElement();
      HeadingElement errorTitle = new HeadingElement.h2();
      errorTitle.text = 'Ошибка подготовки штрихкода для \"$barcodeValue\"';
      String barcodeInfoText = barcodesInfo[barcode.toUpperCase()];
      if(barcodeInfoText == null) {
        barcodeInfoText = 'Выбранный  метод кодирования <b>${barcode.toUpperCase()}</b> не поддерживается';
      }
      barcodeInfo.appendHtml(barcodeInfoText);
      error.append(errorTitle);
      error.append(barcodeInfo);
      error.style.display = 'block';
      querySelector('#success').style.display = 'none';
    }
  }
}
